File: templating/src/main/python/jsonteng/rules/rule_map.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

from collections import OrderedDict

from .at_rule import AtRule
from .for_each_rule import ForEachRule
from .exists_rule import ExistsRule
from .len_rule import LenRule
from .one_of_rule import OneOfRule


_rule_class_map = {
    AtRule.name: AtRule,
    ExistsRule.name: ExistsRule,
    ForEachRule.name: ForEachRule,
    LenRule.name: LenRule,
    OneOfRule.name: OneOfRule
}


def get_rule_map(rule_resolver):
    """
    Return a rule map of known rules.
    :param rule_resolver: Rule resolved used by rules.
    :type rule_resolver: 'RuleResolver'
    :return: A map of rules.
    :rtype: 'dict'
    """
    rule_map = OrderedDict()
    for rule_name, rule in _rule_class_map.items():
        rule_map[rule_name] = rule(rule_resolver)
    return rule_map


def add_rule(rule_name, rule):
    """
    Add a rule to the rule collection.
    :param rule_name: Rule name.
    :type rule_name: 'str'
    :param rule: Rule class.
    :type rule: 'RuleBase'
    :return:
    """
    _rule_class_map[rule_name] = rule


def get_rule_names():
    """
    Get all rule names.
    :return: Rule names.
    :rtype: 'list'
    """
    return _rule_class_map.keys()

File: templating/core/src/main/java/com/vmware/jsonteng/rules/RuleMap.java
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package com.vmware.jsonteng.rules;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.Map;

import com.vmware.jsonteng.RuleResolver;
import com.vmware.jsonteng.TemplateEngineException;

public class RuleMap {
    private static final Map<String, Class<?>> ruleClassMap;

    static {
        ruleClassMap = new Hashtable<>();

        ruleClassMap.put(AtRule.name, AtRule.class);
        ruleClassMap.put(ExistsRule.name, ExistsRule.class);
        ruleClassMap.put(ForEachRule.name, ForEachRule.class);
        ruleClassMap.put(LenRule.name, LenRule.class);
        ruleClassMap.put(OneOfRule.name, OneOfRule.class);
    }

    public static Map<String, RuleBase> getRuleMap(RuleResolver ruleResolver) throws TemplateEngineException {
        Map<String, RuleBase> ruleMap = new HashMap<>();
        for (Map.Entry<String, Class<?>> entry : ruleClassMap.entrySet()) {
            try {
                Constructor<RuleBase> ctor = (Constructor<RuleBase>) entry.getValue().getConstructor(
                        RuleResolver.class);
                RuleBase object = ctor.newInstance(ruleResolver);
                ruleMap.put(entry.getKey(), object);
            } catch (NoSuchMethodException | InstantiationException |
                    IllegalAccessException | InvocationTargetException e) {
                throw new TemplateEngineException("Failed to add rule " + entry.getKey(), e);
            }
        }
        return ruleMap;
    }

    public static void addRule(String ruleName, Class<?> rule) {
        ruleClassMap.put(ruleName, rule);
    }

    public static String[] getRuleNames() {
        return ruleClassMap.keySet().toArray(new String[0]);
    }
}
